src/mixer/basetrackplayer.cpp

```diff
@@ -256,8 +269,9 @@ TrackPointer BaseTrackPlayerImpl::unloadTrack() {
         QListIterator<CuePointer> it(cuePoints);
         while (it.hasNext()) {
             CuePointer pCue(it.next());
-            if (pCue->getType() == mixxx::CueType::Loop) {
+            if (pCue->getType() == mixxx::CueType::Loop && pCue->getHotCue() == Cue::kNoHotCue) {
```

Comment:
This will prevent saving adjustments to loops imported from Serato in 2.3.

Reply:
Yes, and that makes sense. 2.3 does not have support for saved loops after all, and we shouldn't modify the imported saved loops silently. I can't use my saved loops in 2.4 if 2.3 already overwrite them on import. 

If the track only has loops *with* a hotcue number, we still load a loop. When the track is unloaded, Mixxx will add a new loop cue without a hotcue number instead of overwriting the saved loop. Next time you load this track, the cue without hotcue number should be preferred.

Reply:
The user should be able to save modifications to the single loop that was loaded before.

> Mixxx will add a new loop cue without a hotcue number instead of overwriting the saved loop.

I think this would be confusing. Users would end up with duplicated loops.

Reply:
Yes, they end up with exactly one additional loop that might or might now be the same as a saved loop. If we start changing the saved loops, user *mustn't* set a new beatloop or they'll lose their first saved loop without them noticing until 2.4.

Reply:
Why would they lose their saved loop? That would only happen if we remove the hotcue number. I don't think we need to remove the hotcue number if we're not using it yet in 2.3.

Reply:
1. In Serato, I set a 4-beat loop at position X and save it at loop index 1.
2. I import the track into Mixxx 2.3 and the loop is loaded as "restored loop".
3. I let the track play. At position Y I press the loop button on my controller that sets an 8-beat loop.
4. I unload the track. My saved loop is now overwritten with that new loop.
5. I upgrade to Mixxx 2.4 which has saved loop support. Unfortunately my saved 4-beat loop at position X is gone.

Saving changes to a saved loop should be an explicit action. Since we don't have support for that yet, but we should not silently overwrite the saved loops. We we do that, the only way to preserve your saved loops is not set any new loops at all, only toggle the restored loop.
Also, this is completely unobvious from the user perspective - I woundn't expect to lose my saved loops just because I hit the beatloop button.

Reply:
Maybe we should just import one loop for 2.3 and advise users to reimport from Serato tags when 2.4 is released.

Reply:
That information shouldn't disappear unless the user does something unusual like intentionally delete the Serato tags, right?

Reply:
In contrast, this implementation creates a new loop cue without hotcue number if no existing loop cue without hotcue number is found. Next time you load that track, it will precedence over the saved loops (that have a hotcue number).

Reply:
I still do not understand how you would lose a saved loop. If you modify the loop, you modify it. That should not change the hotcue number. I think silently duplicating a loop would be worse. Then in 2.4 the user will probably have two overlapping loops.

Reply:
> I still do not understand how you would lose a saved loop. If you modify the loop, you modify it.

I wouldn't even know that I modified a loop. To the user, that loop just looks like it's a volatile, restored loop.

> I think silently duplicating a loop would be worse. Then in 2.4 the user will probably have two overlapping loops.

Why is that worse? Can you elaborate what the problem with that is?

**Before:** Every track has at most 1 loop cue (the restored loop, i.e. the most recent loop you used for this track).

**After:** Every track has at most n+1 loop cues (the restored loop, i.e. the most recent loop you used for this track + n saved loops you imported from Serato).

When unloading a track, restored loop (without a hotcue number) is overwritten in both cases to always reflect the most recently used loop for the track.

Let's take a step back and reiterate what the purpose of Serato loop import for 2.3 is:
- We do not have Saved Loop support in 2.3, but all the data structures to handle them are there.
- Hence, we can import Serato's saved loops in 2.3, so that after upgrading to 2.4, users do not need to reimport the metadata (and lose all hotcue changes they made in Mixxx).
- Changing/using saved loops is not possible in Mixxx, so saved loops just exist in the DB but are not accessible. Again, we only do this to simplify upgrading to 2.4, so that users can start using their saved loops from Serato right away.
- As a small bonus: If you load the track for the first time and no restored most recent loop exists but loops were imported from Serato, it restores the first saved loop for Serato instead.

Reply:
> Why is that worse? Can you elaborate what the problem with that is?

I think it would be strange and confusing, but maybe it won't actually cause problems beyond being surprising. If the user activates the overlapped saved loop in 2.4, that will become the active loop and override the old unnumbered loop.

Reply:
I think I understand what you mean now about losing a saved loop. If Mixxx imports a saved loop, then the user sets a loop at a totally different location, it probably would not be desired for that new loop to overwrite the imported loop.

Reply:
So I agree with you that copying the imported loop is probably the least bad option for 2.3.

Reply:
However it doesn't seem your current implementation does make a copy of the imported loop and remove its hotcue number. I think the point when the loop is loaded is the wrong place to handle this. How about doing that in the Serato and Rekordbox import code?

Reply:
ping

Reply:
Sorry, looks like I missed that comment.

> However it doesn't seem your current implementation does make a copy of the imported loop and remove its hotcue number.

I don't understand. There is no need to make a copy of the imported loop cue and remove its hotcue number, we just select the loop cue we want to restore, take the position and length and set up a loop for that track region. When the track is unloaded, we just detect the currently used loop position/length and save it to a loop cue without a hotcue number (either overwrite the existing one or create a new one if none exists).

IMHO this is the appropriate thing to do and should mirror exactly what the current implementation is supposed to do when taking numbered loop cues into account. Except for the bugfix (original code only expected unnumbered loop cues), the only actual change of behaviour here is the possibility to fall back to a numbered loop on track load if no unnumbered loop exists.

> I think the point when the loop is loaded is the wrong place to handle this. How about doing that in the Serato and Rekordbox import code? Here where Mixxx loads the loop we would only load a loop without a hotcue number for 2.3.

I think it is the the correct place. The serato parser should not have to know anything about how restored loops are handled in Mixxx IMHO, just return the cues and loops from the metadata.

All the code to backup/restore loops between sessions is in these two functions and we should not spread it to other parts of the code base.